activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -911,28 +911,96 @@ def add_index_options(table_name, column_name, options = {}) #:nodoc:
       end
 
       protected
+
+        def add_index_length(option_strings, column_names, options = {})
+          if options.is_a?(Hash) && length = options[:length]
+            case length
```

Comment:
Agree, I should also check if options can ever not be a hash, otherwise the initial check could be left out. I extracted this method from the mysql adapter and than added additional methods in the same style.